cnxarchive/views/xpath.py

```diff
@@ -0,0 +1,197 @@
+#        Helpers       #
+# #################### #
+
+
+def xpath_book(request, uuid, version):
+    """
+    Given a request, book UUID and version:
+
+    if Accept: 'application/xhtml+xml' is in the headers, returns the HTML tree
+    of the book, with each module title a link which returns the xpath query
+    result for that page UUID (and version if supplied, else uses most recent),
+    or
+
+    otherwise, returns a JSON object of results, each result containing:
+    module_name,
+    module_uuid,
+    xpath_results, an array of strings, each an individual xpath result.
+    """
+
+    if 'application/xhtml+xml' in request.headers.get('ACCEPT', ''):
+        return xpath_book_html(request, uuid, version)
+    else:
+        return execute_xpath(request, 'xpath', uuid, version)
+
+
+def xpath_page(request, uuid, version):
+    """Given a page UUID (and optional version), returns a JSON object of
+    results, as in xpath_book()"""
+    return execute_xpath(request, 'xpath-module', uuid, version)
+
+
+def _get_content_json_xpath(request, uuid, version):
+    """Return a content as a dict from its ident-hash (uuid@version)."""
+
+    settings = get_current_registry().settings
+    as_collated = asbool(request.GET.get('as_collated', True))
+    ident_hash = join_ident_hash(uuid, version)
+    with psycopg2.connect(settings[config.CONNECTION_STRING]) as db_connection:
+        with db_connection.cursor() as cursor:
+            result = get_content_metadata(uuid, version, cursor)
+            # Grab the collection tree.
+            result['tree'] = get_tree(ident_hash, cursor,
+                                      as_collated=as_collated)
+            result['collated'] = as_collated
+            if not result['tree']:
+                # If collated tree is not available, get the uncollated
+                # tree.
+                result['tree'] = get_tree(ident_hash, cursor)
+                result['collated'] = False
+
+    return result
+
+
+def xpath_book_html(request, uuid, version):
+    """Build the HTML tree."""
+    result = _get_content_json_xpath(request, uuid, version)
+
+    content = tree_to_html_xpath(result['tree'])
+
+    resp = request.response
+    resp.status = "200 OK"
+    resp.content_type = 'application/xhtml+xml'
+    resp.body = content
+    return resp
+
+
+def tree_to_html_xpath(tree):
+    """Return html list version of book tree."""
+    ul = etree.Element('ul')
+    html_listify_xpath([tree], ul)
+    return HTML_WRAPPER.format(etree.tostring(ul))
+
+
+def html_listify_xpath(tree, root_ul_element, parent_id=None):
+    """Recursively construct HTML nested list version of book tree, adding
+    hrefs to the page titles which are links to the the API endpoint which runs
+    the xpath on that page CNXML. The original caller should not call this
+    function with the `parent_id` defined.
+
+    """
+    request = get_current_request()
+    querystring = "&q=" + request.params.get('q', '')
```

Comment:
Is it possible to use `urllib.urlencode` instead of hand building the query string?  See https://docs.python.org/2.7/library/urllib.html#examples